### DISCUSSION TITLE
Horizon Quirks?

### DISCUSSION THREAD

**Discussion by r0w4n** [XCSoar/XCSoar#1212] [General]
Hello,

I've been playing around with an IMU and utilising the Levil sentence to get the IMU data into xcsoar. Everything seems to be working great. I haven't flown with it and compared it with the primary AHRS (LXNAV S80) which I plan to do.

The thing I want to discuss here is the way that XCSoar appears to work against the "Levil" data. It might be that I'm doing something wrong and that's why this is a discussion rather than a bug report.

At a pitch up of >= 50° XCSoar's horizon switches from being in the blue "sky" to pointing to the gound
At a roll of >= +/- 90° XCSoar's horizon stops rolling and even though the roll angle continues to increase XCSoar continues to display the roll as a 90° angle

Thoughts? Are any of these by desgin? Anyone else able to confirm that this is the behaviour that they experience?

**Comment by MaxKellermann** [comment]
In your setup, XCSoar isn't using any AHRS data because XCSoar's LX driver does not have any support for AHRS features.

**Comment by r0w4n** [comment]
I think I confused the conversation by mentioning the lxnav s80 as a visual test reference to see how well the imu performs. 

In this scenario I'm using an imu chip to provide roll, pitch and yaw data to XCSoar in the levil sentence format. The lxnav isn't connected.

XCSoar is displaying the horizon data perfectly until either a pitch up of 50 degrees or a roll of greater than 90 degrees. 
I've videoed this behaviour, hopefully it shows this behaviour clearly.

https://user-images.githubusercontent.com/6553343/236665534-68e2a2af-1677-4fdf-8536-3128994fe904.mp4

**Comment by r0w4n** [comment]
Looking at the code, it looks like for some reason the the roll and pitch have been restricted to as I've described.

The following line appears to restrict the role to < ±90
`auto phi = Clamp(bank_degrees, -89., 89.);` [HorizonRenderer.cpp line 45](https://github.com/XCSoar/XCSoar/blob/bedf78b0f63ed9ae419126bb89e268170023845a/src/Renderer/HorizonRenderer.cpp#L45)

And the following line appears to restrict the pitch to 50
`  auto alpha = Angle::acos(Clamp(pitch_degrees / 50, -1., 1.));` [HorizonRenderer.cpp line 46](https://github.com/XCSoar/XCSoar/blob/bedf78b0f63ed9ae419126bb89e268170023845a/src/Renderer/HorizonRenderer.cpp#LL46C3-L47C44)

It would be great to understand the decision to do these restrictions.

**Comment by r0w4n** [comment]
https://github.com/XCSoar/XCSoar/pull/1220
